In [1]:
import numpy as np
import pandas as pd
import pyodbc
import sqlalchemy
import sqlite3

%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [2]:
%sql sqlite://

'Connected: @None'

## 1. SELECT Single & multiple columns

While SQL can be used to create and modify databases, the focus of this course will be querying databases. A query is a request for data from a database table (or combination of tables). Querying is an essential skill for a data scientist, since the data you need for your analyses will often live in databases.

In [3]:
from subprocess import check_output

conn = sqlite3.connect('/kaggle/input/sqlite-sample/SQL_test.sqlite')
exectue = conn.cursor()

### 1.1. Selecting single column

In `SQL`, we can select data from a table using a **SELECT** `statement`. 

For example, the following query selects the `A09_village` column from the `Farms` table

In [4]:
SQL_farms_single_col = pd.read_sql(
    """
    SELECT A09_village 
    FROM Farms
    """, con = conn)
SQL_farms_single_col

A09_village
0           God
1           God
2           God
3           God
4           God
..          ...
345      Ndedja
346      Ndedja
347      Ndedja
348      Ndedja
349      Ndedja

[350 rows x 1 columns]

### 1.2. Selecting multiple columns

Sometimes, you may want to select all columns from a table. Typing out every column name would be a pain, so there's a handy shortcut:

In [5]:
SQL_farms_single_col = pd.read_sql(
    """
    SELECT Id, A06_province, A07_district, G02_months_lack_food, A01_interview_date
    FROM Farms
    """, con = conn)
SQL_farms_single_col

Id A06_province A07_district  \
0      1       Manica       Manica   
1      2       Manica       Manica   
2      3       Manica       Manica   
3      4       Manica       Manica   
4      5       Manica       Manica   
..   ...          ...          ...   
345  346       Sofala   Nhamatanda   
346  347       Sofala   Nhamatanda   
347  348       Sofala   Nhamatanda   
348  349       Sofala   Nhamatanda   
349  350       Sofala   Nhamatanda   

                           G02_months_lack_food A01_interview_date  
0                                       ['Jan']         17/11/2016  
1          ['Jan', 'Sept', 'Oct', 'Nov', 'Dec']         17/11/2016  
2    ['Jan', 'Feb', 'Mar', 'Oct', 'Nov', 'Dec']         17/11/2016  
3                 ['Sept', 'Oct', 'Nov', 'Dec']         17/11/2016  
4                 ['Aug', 'Sept', 'Oct', 'Nov']         17/11/2016  
..                                          ...                ...  
345                ['Jan', 'Feb', 'Mar', 'Dec']         22/08/2017  
346                ['Jan', 'Feb', 'Mar', 'Dec']         22/08/2017  
347                                    ['none']         22/08/2017  
348                ['Jan', 'Feb', 'Mar', 'Dec']         22/08/2017  
349                ['Jan', 'Feb', 'Mar', 'Dec']         22/08/2017  

[350 rows x 5 columns]

## 2. SELECTING ALL COLUMNS
In the real world, you will often want to select multiple columns. Luckily, SQL makes this really easy. To select multiple columns from a table, simply separate the column names with commas!

For example, this query selects two columns, name and birthdate, from the people table:
### 2.1. All columns & all rows

In [6]:
SQL_farms_all_cols = pd.read_sql(
    """
    SELECT * FROM Farms
    """, con = conn)
SQL_farms_all_cols

Id Country A01_interview_date  A03_quest_no                 A04_start  \
0      1     Moz         17/11/2016             1  2017-03-23T09:49:57.000Z   
1      2     Moz         17/11/2016             1  2017-04-02T09:48:16.000Z   
2      3     Moz         17/11/2016             3  2017-04-02T14:35:26.000Z   
3      4     Moz         17/11/2016             4  2017-04-02T14:55:18.000Z   
4      5     Moz         17/11/2016             5  2017-04-02T15:10:35.000Z   
..   ...     ...                ...           ...                       ...   
345  346     Moz         22/08/2017           207  2017-08-22T10:35:17.000Z   
346  347     Moz         22/08/2017           209  2017-08-22T10:42:15.000Z   
347  348     Moz         22/08/2017           205  2017-08-23T05:08:13.000Z   
348  349     Moz         22/08/2017           206  2017-08-23T05:39:40.000Z   
349  350     Moz         22/08/2017           208  2017-08-23T05:53:59.000Z   

                      A05_end A06_province A07_district A08_ward A09_village  \
0    2017-04-02T17:29:08.000Z       Manica       Manica  Bandula         God   
1    2017-04-02T17:26:19.000Z       Manica       Manica  Bandula         God   
2    2017-04-02T17:26:53.000Z       Manica       Manica  Bandula         God   
3    2017-04-02T17:27:16.000Z       Manica       Manica  Bandula         God   
4    2017-04-02T17:27:35.000Z       Manica       Manica  Bandula         God   
..                        ...          ...          ...      ...         ...   
345  2017-08-22T10:41:35.000Z       Sofala   Nhamatanda   Lamego      Ndedja   
346  2017-08-23T06:21:54.000Z       Sofala   Nhamatanda   Lamego      Ndedja   
347  2017-08-23T05:27:06.000Z       Sofala   Nhamatanda   Lamego      Ndedja   
348  2017-08-23T05:53:54.000Z       Sofala   Nhamatanda   Lamego      Ndedja   
349  2017-08-23T06:04:39.000Z       Sofala   Nhamatanda   Lamego      Ndedja   

     ...  F13_du_look_aftr_cows  \
0    ...                     no   
1    ...                     no   
2    ...                     no   
3    ...                     no   
4    ...                     no   
..   ...                    ...   
345  ...                     no   
346  ...                     no   
347  ...                     no   
348  ...                     no   
349  ...                     no   

                                       F14_items_owned  G01_no_meals  \
0    ['bicycle', 'television', 'solar_panel', 'table']             2   
1    ['cow_cart', 'bicycle', 'radio', 'cow_plough',...             2   
2                                      ['solar_torch']             2   
3    ['bicycle', 'radio', 'cow_plough', 'solar_pane...             2   
4    ['motorcyle', 'radio', 'cow_plough', 'mobile_p...             2   
..                                                 ...           ...   
345                                          ['radio']             3   
346                                               None             3   
347  ['bicycle', 'radio', 'solar_torch', 'mobile_ph...             3   
348                                          ['radio']             3   
349         ['bicycle', 'solar_panel', 'mobile_phone']             3   

                           G02_months_lack_food  \
0                                       ['Jan']   
1          ['Jan', 'Sept', 'Oct', 'Nov', 'Dec']   
2    ['Jan', 'Feb', 'Mar', 'Oct', 'Nov', 'Dec']   
3                 ['Sept', 'Oct', 'Nov', 'Dec']   
4                 ['Aug', 'Sept', 'Oct', 'Nov']   
..                                          ...   
345                ['Jan', 'Feb', 'Mar', 'Dec']   
346                ['Jan', 'Feb', 'Mar', 'Dec']   
347                                    ['none']   
348                ['Jan', 'Feb', 'Mar', 'Dec']   
349                ['Jan', 'Feb', 'Mar', 'Dec']   

                                G03_no_food_mitigation  gps:Latitude  \
0    ['na', 'rely_less_food', 'reduce_meals', 'day_...    -19.112259   
1    ['na', 'reduce_meals', 'restrict_adults', 'bor..

### 2.2. LIMITS rows

And if we want to return a certain numbers of rows; we can add a keyword **LIMIT** to limit the numbers of results returned!

In [7]:
SQL_crops_6_rows = pd.read_sql(
    """
    SELECT * FROM Farms
    LIMIT 6;
    """, con = conn)
SQL_crops_6_rows

Id Country A01_interview_date  A03_quest_no                 A04_start  \
0   1     Moz         17/11/2016             1  2017-03-23T09:49:57.000Z   
1   2     Moz         17/11/2016             1  2017-04-02T09:48:16.000Z   
2   3     Moz         17/11/2016             3  2017-04-02T14:35:26.000Z   
3   4     Moz         17/11/2016             4  2017-04-02T14:55:18.000Z   
4   5     Moz         17/11/2016             5  2017-04-02T15:10:35.000Z   
5   6     Moz         17/11/2016             6  2017-04-02T15:27:25.000Z   

                    A05_end A06_province A07_district A08_ward A09_village  \
0  2017-04-02T17:29:08.000Z       Manica       Manica  Bandula         God   
1  2017-04-02T17:26:19.000Z       Manica       Manica  Bandula         God   
2  2017-04-02T17:26:53.000Z       Manica       Manica  Bandula         God   
3  2017-04-02T17:27:16.000Z       Manica       Manica  Bandula         God   
4  2017-04-02T17:27:35.000Z       Manica       Manica  Bandula         God   
5  2017-04-02T17:28:02.000Z       Manica       Manica  Bandula         God   

   ...  F13_du_look_aftr_cows  \
0  ...                     no   
1  ...                     no   
2  ...                     no   
3  ...                     no   
4  ...                     no   
5  ...                     no   

                                     F14_items_owned  G01_no_meals  \
0  ['bicycle', 'television', 'solar_panel', 'table']             2   
1  ['cow_cart', 'bicycle', 'radio', 'cow_plough',...             2   
2                                    ['solar_torch']             2   
3  ['bicycle', 'radio', 'cow_plough', 'solar_pane...             2   
4  ['motorcyle', 'radio', 'cow_plough', 'mobile_p...             2   
5                                               None             2   

                         G02_months_lack_food  \
0                                     ['Jan']   
1        ['Jan', 'Sept', 'Oct', 'Nov', 'Dec']   
2  ['Jan', 'Feb', 'Mar', 'Oct', 'Nov', 'Dec']   
3               ['Sept', 'Oct', 'Nov', 'Dec']   
4               ['Aug', 'Sept', 'Oct', 'Nov']   
5                      ['Aug', 'Sept', 'Oct']   

                              G03_no_food_mitigation  gps:Latitude  \
0  ['na', 'rely_less_food', 'reduce_meals', 'day_...    -19.112259   
1  ['na', 'reduce_meals', 'restrict_adults', 'bor...    -19.112477   
2           ['na', 'restrict_adults', 'lab_ex_food']    -19.112108   
3  ['na', 'reduce_meals', 'restrict_adults', 'lab...    -19.112229   
4                     ['na', 'go_forest', 'migrate']    -19.112217   
5  ['borrow_food', 'lab_ex_food', 'seek_government']    -19.112196   

  gps:Longitude gps:Altitude gps:Accuracy  \
0     33.483456        698.0         14.0   
1     33.483416        690.0         19.0   
2     33.483450        674.0         13.0   
3     33.483424        679.0          5.0   
4     33.483425        689.0         10.0   
5     33.483392        692.0         12.0   

                                  instanceID  
0  uuid:ec241f2c-0609-46ed-b5e8-fe575f6cefef  
1  uuid:099de9c9-3e5e-427b-8452-26250e840d6e  
2  uuid:193d7daf-9582-409b-bf09-027dd36f9007  
3  uuid:148d1105-778a-4755-aa71-281eadd4a973  
4  uuid:2c867811-9696-4966-9866-f35c3e97d02d  
5  uuid:daa56c91-c8e3-44c3-a663-af6a49a2ca70  

[6 rows x 61 columns]

## 3. SELECT DISTINCT

Often your results will include many duplicate values. If you want to select all the unique values from a column, you can use the DISTINCT keyword.

This might be useful if, for example, you're interested in knowing which languages are represented in the films table:


In [8]:
SQL_crops_distinct = pd.read_sql(
    """
    SELECT DISTINCT A01_interview_date
    FROM Farms
    """, con = conn)
SQL_crops_distinct

A01_interview_date
0          17/11/2016
1          16/11/2016
2          16/12/2016
3          21/11/2016
4          24/11/2016
5          18/11/2016
6          23/11/2016
7          25/11/2016
8          28/11/2016
9          26/04/2017
10         27/04/2017
11         28/04/2017
12         30/04/2017
13         03/05/2017
14         04/05/2017
15         11/05/2017
16         18/05/2017
17         03/06/2017
18         04/06/2017
19         11/04/2017
20         12/04/2017
21         17/06/2017
22         22/06/2017
23         23/06/2017
24         24/06/2017
25         25/06/2017
26         26/06/2017
27         27/06/2017
28         28/06/2017
29         29/06/2017
30         30/06/2017
31         01/07/2017
32         02/07/2017
33         04/07/2017
34         05/07/2017
35         06/07/2017
36         07/07/2017
37         09/07/2017
38         10/07/2017
39         13/07/2017
40         14/07/2017
41         16/07/2017
42         17/07/2017
43         18/07/2017
44         19/07/2017
45         20/07/2017
46         24/07/2017
47         25/07/2017
48         01/12/2016
49         21/08/2017
50         22/08/2017

We can see that there are 349 observations in the database `Farms` but the `unique_values` of `A01_interview_date 's column` has 51 `distinct_values` only.